### DISCUSSION TITLE
Question: Regular Markdown HTML syntax ignored

### DISCUSSION THREAD

**Discussion by ovflowd** [mdx-js/mdx#2231] [Q&A] [answered]
Hey there 👋 I was wondering to get an answer on the following, I recall (when I used Gatsby) that regular Markdown `.md` files could still have HTML syntax such as `<div class="something">` and it would still allow me to do something like `<SomeJSXComponent />` if I passed on `<MDXPRovider components={{ SomeJSXComponent: (something) => <div>hello there</div> }} />`.

Of course, I'm assuming this based on my memory, but I just wanted to double-check if this is something actually supported on MDXv2, as I recall, it worked on MDXv1.

**Comment by wooorm** [comment, marked as answer]
Hey!

- Depends what you mean with “work”. Lots of things crashed in MDX 1 normally for various syntax reasons because strings were passed through. Gatsby had a rather complex system to parse those things again. Which wasn’t part of the MDX projects but their own language. That’s no longer needed in MDX 2, because MDX does all of that.
- Presumably, `<SomeJSXComponent />` is just an example, but that isn’t valid HTML. That’s JSX. So if you want to type things like that, use the MDX language itself.
- It is possible to use markdown, and also to use HTML in markdown, with MDX. I don’t think it’s a smart idea though. You’re introducing *another* parser. A lot of work. A lot of time. Because you want to pass strings of HTML through to React. So we need to parse HTML to generate react calls so that react can build a virtual dom so that it can generate a DOM which is equivalent to the HTML. Here’s how to do it:

https://github.com/mdx-js/mdx/blob/f77c33f5b21b6c19eebdb646572d7dc34a9dd875/packages/mdx/test/compile.js#L768

**Reply by ovflowd** [reply]
@wooorm yeah, this is not ideal, what I'm referring is to this example: https://raw.githubusercontent.com/ovflowd/nodejs.org/feat/nextjs-infrastructure/pages/en/docs/index.md

Definitely, in the future, we want to avoid doing this, but temporarily we need to support HTML within Markdown. Before this migration we're doing we've used (https://github.com/metalsmith/markdown) to parse Markdown into HTML.

What I kinda want to accomplish here, is that on regular Markdown files MDX is still able to parse HTML into JSX. (I was hoping it would consider `<SomeJSXComponent />` as just a regular HTML tag and then forward it to the `MDXPRovider` `components` table.

**Reply by ovflowd** [reply]
Is there a way we could accomplish this without extra complications? Otherwise the alternative is to convert the file I mentioned above (/en/docs/index.md) into MDX.

**Reply by wooorm** [reply]
That syntax you show isn’t HTML. That doesn’t work in an HTML parser.
I can give you JSX and I can give you HTML, but this is frankensteins monster

**Reply by wooorm** [reply]
Did metalsmith parse this? Or did Gatsby?

**Reply by ovflowd** [reply]
Metalsmith did parse that file. Gatsby is used in another repository, but we did similar things. (Having a JSX call within a regular .md file), for example, [here](https://raw.githubusercontent.com/nodejs/nodejs.dev/main/content/api/v19/cli.en.md)

**Reply by ovflowd** [reply]
> That syntax you show isn’t HTML. That doesn’t work in an HTML parser.
I can give you JSX and I can give you HTML, but this is frankensteins monster

I have no idea then how Gatsby and Metalsmith are able to do this 😞 (And Gatsby uses MDXv1 but with modifications I guess)

**Reply by wooorm** [reply]
I specifically mean: `<NodeApiVersionLinks />` btw.

> Metalsmith did parse that file

Did it do more that parse? (sorry, I should’ve been more clear). How and why do you combine Metalsmith with Gatsby?

> I have no idea then how Gatsby and Metalsmith are able to do this

You have to clarify what *this* means.
Because MDX can parse that file, if you call it `.mdx`, or if you set `format: 'mdx'`. Because that looks like MDX.
And MDX can also parse that file as markdown (`.md`), and handle the HTML in markdown, as I showed above, but `<NodeApiVersionLinks />` is not HTML. This is the equivalent in HTML:

```
document.body.innerHTML = `<div class="highlight-box">
<h3>xxx</h3>
<NodeApiVersionLinks />
<p><a href="https://nodejs.org/docs/">All versions</a></p>
</div>
`; document.body.innerHTML
```
```html
"<div class=\"highlight-box\">
<h3>xxx</h3>
<nodeapiversionlinks>
<p><a href=\"https://nodejs.org/docs/\">All versions</a></p>
</nodeapiversionlinks></div>
"
```

---

But now I no longer know what you want. You asked about HTML, but you show JSX. MDX can parse JSX. And it can parse HTML?

**Reply by ovflowd** [reply]
> Did it do more that parse? (sorry, I should’ve been more clear). How and why do you combine Metalsmith with Gatsby?

We are talking about two different things. The Nodejs.org repository uses Metalsmith. The Nodejs.dev repository uses Gatsby. I'm saying that the `docs/index.md` was parsed from Markdown to HTML fine by Metalsmith. And that `/api/` Markdown file was parsed fine by Gatsby.

> But now I no longer know what you want. You asked about HTML, but you show JSX. MDX can parse JSX. And it can parse HTML?

I'm getting confused. `<Something />` is 100% valid HTML. If you append this to the DOM it gets translated into `<something></something>`. I thought that MDX would simply map Markdown tags such as `###` into their respective HTML tags `h3` and then look into a match on the `MDXProvider components={}` table.

That's why I came up with `<SomeComponent />` as a "custom" HTML tag that MDX would just try to find a match on the `components` table 😅

**Reply by wooorm** [reply]
> is 100% valid HTML

Anything is 100% valid HTML: HTML has no syntax errors (the BNF grammar for HTML, and markdown, is `.*`). But what you show doesn’t do what you expect. Your `<something></something>` understanding is incorrect. I showed how a browser sees it in my previous comment. Again:

```html
<div class="highlight-box">
<h3>xxx</h3>
<nodeapiversionlinks>
<p><a href="https://nodejs.org/docs/">All versions</a></p>
</nodeapiversionlinks></div>
```

It’s an unclosed element, it runs until the next parent is closed. That is not what you want.

---

> I thought that MDX would simply map Markdown tags such as ### into their respective HTML tags h3 and then look into a match on the MDXProvider components={} table.

MDX does several things, component swapping too. But it’s not simple. It’s quiet complex, how to parse markdown and JavaScript (and HTML) together.

**Reply by wooorm** [reply]
> `docs/index.md`, `/api/`

You only linked `cli.en.md` as far as I can see, the other one 404s. Which one is `/api/`?

**Reply by ovflowd** [reply]
> It’s an unclosed element, it runs until the next parent is closed. That is not what you want.

I thought that's where MDXProvider kicks in. Again, if I have a `###` or `h3` the MDXProvider will map the h3 tag to whatever I have inside the `components` list for `h3` even for `format: 'md'`.

Hence *why* having MDXProvider mapping the `<NodeApiVersonLinks>` into the component from the table.

> MDX does several things, component swapping too. But it’s not simple. It’s quiet complex, how to parse markdown and JavaScript (and HTML) together.

I thought that's what MDX does by default, parsing both HTML and Markdown. Because HTML tags are not part of the MD spec... 

I'm making a lot of assumptions here 😅

**Reply by ovflowd** [reply]
> You only linked cli.en.md as far as I can see, the other one 404s. Which one is /api/?

Sorry, with /api/ I meant the cli.md which one is giving 404?

**Reply by ovflowd** [reply]
(I've updated my previous comments as I rushed with them 🫠)

**Reply by wooorm** [reply]
Okay, so first, your understanding of providers is incorrect.
Providers are not needed. Providers are ±10 lines of code to hook a specific React context based API into MDX.
You can swap components in without them.
You don’t need providers.
They don’t interact with parsers (they run on the client). They can’t somehow change how an HTML parser sees the XML you write.
Please read through the article on how to use MDX, which explains how to pass components, what providers are, and when you should or should not use them: https://mdxjs.com/docs/using-mdx/

> Again, if I have a `###` or h3 the MDXProvider will map the h3 tag to whatever I have inside the components list for h3 even for format: 'md'.

You mention two things, and there are two formats:

* In MDX, `###` can be swapped out with a custom component. This works.
* In MDX, `<h3>` cannot be swapped out with a custom component, which is by design: type `<H3>` instead, you can pass a component for that (this is your `NodeApiVersonLinks` example, which works)
* In markdown, `###` can be swapped out with a custom component. This works.
* In markdown, `<h3>` can be swapped out with a custom component, if you configure an HTML parser. I showed in my first comment how you can do that with one line of code.

> I thought that's what MDX does by default, parsing both HTML and Markdown. Because HTML tags are not part of the MD spec...

This needs an important note: **how to parse markdown, including how to parse the HTML in markdown**, is defined in the markdown spec. Please see the markdown spec: https://spec.commonmark.org/0.30/. (Also see this, which I recently wrote about how the grammar has to be parsed: https://github.com/wooorm/markdown-rs/blob/main/src/construct/html_flow.rs)

The markdown spec defines a syntax that treats strings of (broken?) XML as a black box. It might be wrong! It might be right. It sees `<div>`? Then you have to output that.
Which is fine if you compile to a string of HTML.
Which is not what this project does.

> If that's the case, I know you mentioned there's a way to do what I want, but if this is hacky or costly way (performance/memory) then I prefer not doing it and searc for another way 😅

Well. You show an MDX file. That includes `<SomeJSXComponent />`. **Just call your file `.mdx` and it works**.

If you actually had HTML, then what I show works, albeit slow. But `<SomeJSXComponent />` is not HTML (in the way you intend it, because it’s not closed and runs very far into the future).  If you want HTML, you *should* use custom elements (with a `-`, so: `<some-custom-element />`), and you  must close them: `<some-custom-element></some-custom-element>`.

**Reply by ovflowd** [reply]
> Please read through the article on how to use MDX, which explains how to pass components, what providers are, and when you should or should not use them: https://mdxjs.com/docs/using-mdx/

Oh boii! I had such an erroneous understanding of how MDX works! All based on assumptions made on how Gatsby was doing things 🤦 

> You mention two things, and there are two formats:

That is enlightening, thank you!

> This needs an important note: how to parse markdown, including how to parse the HTML in markdown, is defined in the markdown spec. Please see the markdown spec: https://spec.commonmark.org/0.30/. (Also see this, which I recently wrote about how the grammar has to be parsed: https://github.com/wooorm/markdown-rs/blob/main/src/construct/html_flow.rs)

Thank you for pointing me out to these files :)

> Well. You show an MDX file. That includes <SomeJSXComponent />. Just call your file .mdx and it works.

That would work if it wasn't by the case that many parts of the file have non-complying Markdown syntax 😅 (I'm going to make a PR within nodejs.org to fix all the issues MDX was complaining about)

> If you actually had HTML, then what I show works, albeit slow. But <SomeJSXComponent /> is not HTML (in the way you intend it, because it’s not closed and runs very far into the future). If you want HTML, you should use custom elements (with a -, so: <some-custom-element />), and you must close them: <some-custom-element></some-custom-element>.

Not going to, I simply made the `/en/docs/index.md` as `index.mdx` and fixed the issues that were non-JSX conforming.

This whole discussion was enlightening. Just a note, I don't want to break the pattern/standard, just trying to understand exactly how MDX works for each thing, so I can understand better what to expect.

Thank you, Titus! 💖

**Reply by wooorm** [reply]
Cool, no worries, feel free to ask questions! Good luck :)
